pandas/core/arrays/categorical.py

```diff
@@ -1268,20 +1271,23 @@ def map(self, mapper, na_action=None):
         >>> cat.map({'a': 'first', 'b': 'second'})
         Index(['first', 'second', nan], dtype='object')
         """
-        if na_action is not None:
-            raise NotImplementedError
+        assert callable(mapper) or is_dict_like(mapper)
 
         new_categories = self.categories.map(mapper)
-        try:
-            return self.from_codes(
-                self._codes.copy(), categories=new_categories, ordered=self.ordered
-            )
-        except ValueError:
-            # NA values are represented in self._codes with -1
-            # np.take causes NA values to take final element in new_categories
-            if np.any(self._codes == -1):
-                new_categories = new_categories.insert(len(new_categories), np.nan)
+
+        not_dictlike_and_no_nans = not (is_dict_like(mapper) and np.nan not in mapper)
+
+        if na_action is None and not_dictlike_and_no_nans and np.any(self._codes == -1):
+            na_value = mapper(np.nan) if callable(mapper) else mapper[np.nan]
+            new_categories = new_categories.insert(len(new_categories), na_value)
             return np.take(new_categories, self._codes)
+        elif new_categories.is_unique and not new_categories.hasnans:
+            new_dtype = CategoricalDtype(new_categories, ordered=self.ordered)
+            return self.from_codes(self._codes.copy(), dtype=new_dtype)
+
+        if np.any(self._codes == -1):
```

Comment:
Could you combine this with the first `if` statement above?

```
if np.any(self._codes == -1):
    if na_action is None and not_dictlike_and_no_nans:
         na_value = mapper(np.nan) if callable(mapper) else mapper[np.nan]
    else:
         na_value = np.nan
    return np.take(new_categories, self._codes)
```

Reply:
it's not possible as-is, because there's the problem that `self.categories.map(mapper)` may both be non-unique and/or contain nans.

I've actually spent quite some time trying to make this a simple as possible, but this is actually more complex than it seems because of all the different cases it'll have to cover. But maybe it's possible and I just couldn't find it :-)